x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,144 @@
+        this.securityContext = context;
+        this.environment = environment;
+        this.nodeService = nodeService;
+        this.sslService = sslService;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateEnrollmentTokenRequest request,
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+            if (keyConfig instanceof StoreKeyConfig == false) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration is not configured with a " +
+                        "keystore"));
+                return;
+            }
+            final List<X509Certificate> caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment).stream()
+                .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1).collect(Collectors.toList());
+            if (caCertificates.size() != 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain a " +
+                    "single PrivateKey entry where the associated certificate is a CA certificate"));
+            } else {
+                final TimeValue expiration = TimeValue.timeValueSeconds(ENROLL_API_KEY_EXPIRATION_SEC);
+                final List<RoleDescriptor> roleDescriptors = new ArrayList<>(1);
+                final String[] clusterPrivileges = { "enroll" };
+                final RoleDescriptor roleDescriptor = new RoleDescriptor("create_enrollment_token", clusterPrivileges, null, null);
+                roleDescriptors.add(roleDescriptor);
+                CreateApiKeyRequest apiRequest = new CreateApiKeyRequest(UUIDs.base64UUID(), roleDescriptors, expiration);
+                final String fingerprint = SslUtil.calculateFingerprint(caCertificates.get(0));
+                createEnrolmentToken(fingerprint, apiRequest, listener);
+            }
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private void createEnrolmentToken(String fingerprint, CreateApiKeyRequest apiRequest,
+                                      ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            generator.generateApiKey(securityContext.getAuthentication(), apiRequest,
+                ActionListener.wrap(
+                    CreateApiKeyResponse -> {
+                        try {
+                            final NodeInfo nodeInfo = nodeService.info(false, false, false, false, false, false,
+                                true, false, false, false, false);
+                            final HttpInfo httpInfo = nodeInfo.getInfo(HttpInfo.class);
```

Comment:
nit: we don't check or use httpInfo and nodeInfo, maybe combine these three lines and not create the 2 extra variables ?